gobblin-core/src/main/java/gobblin/source/extractor/extract/kafka/KafkaSource.java

```diff
@@ -303,16 +376,21 @@ private boolean shouldMoveToLatestOffset(KafkaPartition partition, SourceState s
     return this.moveToLatestTopics.contains(partition.getTopicName()) || moveToLatestTopics.contains(ALL_TOPICS);
   }
 
-  private WorkUnit createEmptyWorkUnit(KafkaPartition partition, long previousOffset) {
+  private WorkUnit createEmptyWorkUnit(KafkaPartition partition, long previousOffset,
+      Optional<State> topicSpecificState) {
     Offsets offsets = new Offsets();
     offsets.setEarliestOffset(previousOffset);
     offsets.setLatestOffset(previousOffset);
     offsets.startAtEarliestOffset();
-    return getWorkUnitForTopicPartition(partition, offsets);
+    return getWorkUnitForTopicPartition(partition, offsets, topicSpecificState);
   }
 
-  private WorkUnit getWorkUnitForTopicPartition(KafkaPartition partition, Offsets offsets) {
+  private WorkUnit getWorkUnitForTopicPartition(KafkaPartition partition, Offsets offsets,
+      Optional<State> topicSpecificState) {
     Extract extract = this.createExtract(DEFAULT_TABLE_TYPE, DEFAULT_NAMESPACE_NAME, partition.getTopicName());
+    if (topicSpecificState.isPresent()) {
+      extract.addAll(topicSpecificState.get());
```

Comment:
It seems `workUnit.getProp()` cannot get properties in the extract of the workunit. Even if it can, the writer partition column property will be in both the workunit and the extract, causing conflict. So it's better to add them to the workunit.